software/apps/baremetal/matmul_i16/main.c

```diff
@@ -7,135 +7,46 @@
-      int32_t ii = (int32_t)i;
-      int32_t jj = (int32_t)j;
-      int32_t lin = ((int32_t)aa * bb * ii * jj + aa * bc * ii + ac * bb * jj +
-                     (int32_t)ac * bc) *
-                    n;
-      int32_t qua =
-          (((int32_t)aa * ba * ii + ab * bb * jj + ab * bc + (int32_t)ba * ac) *
-           (n * (n - 1))) /
-          2;
-      int32_t cub = (((int32_t)ab * ba) * (n * (n - 1) * (2 * n - 1))) / 6;
-      int32_t golden = lin + qua + cub;
-      if (matrix[i * num_columns + j] != golden) {
-        return (i + j) == 0 ? -1 : (int)(i * num_columns + j);
-      }
-      matrix[i * num_columns + j] = 0;
-    }
+  // Initialize data
+  if (core_id == 0) {
+    dma_memcpy_blocking(l1_A, l2_A, matrix_M * matrix_N * sizeof(int16_t));
+    dma_memcpy_blocking(l1_B, l2_B, matrix_N * matrix_P * sizeof(int16_t));
   }
-  return 0;
-}
-
-int test_matrix_multiplication(int16_t *__restrict__ A, int16_t *__restrict__ B,
-                               int32_t *__restrict__ C, uint32_t M, uint32_t N,
-                               uint32_t P, uint32_t core_id,
-                               uint32_t num_cores) {
-  int16_t const A_a = 1;
-  int16_t const A_b = 1;
-  int16_t const A_c = -40;
-  int16_t const B_a = 0;
-  int16_t const B_b = 1;
-  int16_t const B_c = 19;
-
-  // Initialize Matrices
-  init_matrix(A, M, N, A_a, A_b, A_c, core_id, num_cores);
-  init_matrix(B, N, P, B_a, B_b, B_c, core_id, num_cores);
-  // Wait at barrier until everyone is ready
   mempool_barrier(num_cores);
-  // Execute function to test.
-  mempool_start_benchmark();
 
+  // Benchmark
+  mempool_start_benchmark();
 #ifdef __XPULPIMG
-  matmul_unrolled_4x2_pincr_asm_parallel_i16_xpulpv2(A, B, C, M, N, P, core_id,
-                                                     num_cores);
+  matmul_unrolled_4x2_pincr_asm_parallel_i16_xpulpv2(
+      l1_A, l1_B, l1_C, matrix_M, matrix_N, matrix_P, core_id, num_cores);
 #else
-  matmul_unrolled_2x2_parallel_i16_rv32im(A, B, C, M, N, P, core_id, num_cores);
+  matmul_unrolled_2x2_parallel_i16_rv32im(l1_A, l1_B, l1_C, matrix_M, matrix_N,
+                                          matrix_P, core_id, num_cores);
 #endif
-
   mempool_stop_benchmark();
-  // Wait at barrier befor checking
   mempool_barrier(num_cores);
-  if (verify_matrix(C, M, P, N, A_a, A_b, A_c, B_a, B_b, B_c, core_id,
-                    num_cores)) {
-    error = 1;
-    return -1;
-  }
-  return 0;
-}
 
-int main() {
-  uint32_t core_id = mempool_get_core_id();
-  uint32_t num_cores = mempool_get_core_count();
-  // Initialize barrier and synchronize
-  mempool_barrier_init(core_id);
-
-  if (core_id == 0) {
-    error = 0;
-  }
-
-  // Test the Matrix multiplication
-  test_matrix_multiplication(matrix_a, matrix_b, matrix_c, matrix_M, matrix_N,
-                             matrix_P, core_id, num_cores);
-  // wait until all cores have finished
+  // Verify results
+  mempool_check_q32(l1_C, l2_C, matrix_M * matrix_P, 0, 0);
```

Comment:
I just want to quickly confirm if the i16 also uses "mempool_check_q32" or if it's a typo.

Reply:
It is correct, because the app has 32b outputs. The q32 suffix was charged to i32.